### PR TITLE
gfx: Add elliptical border radius support

### PR DESCRIPTION
**Pull request by bjwbell**
TODO: Add code for parsing shorthand border-radius e.g. "border-radius: 10px 5% / 20px".

r? @pcwalton

[Review on Reviewable](https://reviewable.io/reviews/servo/servo/7502)

### CODE REVIEW

File: components/gfx/display_list/mod.rs [hunk 1]
```diff
@@ -961,12 +961,41 @@ pub struct BorderDisplayItem {
...
+    fn default() -> Self {
+        let top_left = Size2D::new(Default::default(),
+                                   Default::default());
+        let top_right = Size2D::new(Default::default(),
+                                    Default::default());
+        let bottom_left = Size2D::new(Default::default(),
+                                      Default::default());
+        let bottom_right = Size2D::new(Default::default(),
+                                       Default::default());
+        BorderRadii { top_left: top_left, top_right: top_right, bottom_left: bottom_left, bottom_right: bottom_right }
```
Comment (pcwalton) [inline components/gfx/display_list/mod.rs]: nit: 100 line limit

File: components/gfx/paint_context.rs [hunk 2]
```diff
@@ -466,6 +466,163 @@ impl<'a> PaintContext<'a> {
...
+                         radius: Size2D<AzFloat>,
+                         start_angle: f32,
+                         end_angle: f32) {
+        if PaintContext::is_zero_radius(&radius) {
+            return;
+        }
+
+        // Calculate kappa constant for partial curve. The sign of angle in the
+        // tangent will actually ensure this is negative for a counter clockwise
+        // sweep, so changing signs later isn't needed.
```
Comment (pcwalton) [inline components/gfx/paint_context.rs]: Do you have a reference where I can read about this? (Code you referenced from is OK.)

Comment (bjwbell) [inline components/gfx/paint_context.rs]: http://mxr.mozilla.org/mozilla-central/source/gfx/2d/PathHelpers.h

I'll add the reference too.

File: components/gfx/paint_context.rs [hunk 3]
```diff
@@ -466,6 +466,167 @@ impl<'a> PaintContext<'a> {
...
+
+        path_builder.bezier_curve_to(&cp1, &cp2, &end_point);
+    }
+
+    #[allow(non_snake_case)]
+    fn border_bounds(bounds: &Rect<f32>,
+                     border: &SideOffsets2D<f32>) -> ((Point2D<f32>, Point2D<f32>),
+                                                      (Point2D<f32>, Point2D<f32>),
+                                                      (Point2D<f32>, Point2D<f32>),
+                                                      (Point2D<f32>, Point2D<f32>)) {
```
Comment (pcwalton) [inline components/gfx/paint_context.rs]: Why not return two Rects?

File: components/gfx/paint_context.rs [hunk 4]
```diff
@@ -466,6 +466,167 @@ impl<'a> PaintContext<'a> {
...
+        ((box_TL, inner_TL), (box_TR, inner_TR), (box_BR, inner_BR), (box_BL, inner_BL))
+    }
+
+    #[allow(non_snake_case)]
+    fn corner_bounds(bounds: &Rect<f32>,
+                     border: &SideOffsets2D<f32>,
+                     radii: &BorderRadii<AzFloat>) -> ((Point2D<f32>, Size2D<f32>),
+                                                       (Point2D<f32>, Size2D<f32>),
+                                                       (Point2D<f32>, Size2D<f32>),
+                                                       (Point2D<f32>, Size2D<f32>)) {
```
Comment (pcwalton) [inline components/gfx/paint_context.rs]: Why not a Rect and a SideOffsets2D?

File: components/gfx/paint_context.rs [hunk 5]
```diff
@@ -466,6 +466,167 @@ impl<'a> PaintContext<'a> {
...
+        }
+
+        if PaintContext::is_zero_radius(radius) {
+            return;
+        }
+        let ellipse = Ellipse { origin: *origin, width: radius.width, height: radius.height };
+        let simple_border = compatible_borders_corner(&radius,
+                                                      (outer_border.x - inner_border.x).abs(),
+                                                      (outer_border.y - inner_border.y).abs());
+        let corner_angle = match (simple_border, inner_border.x >= outer_border.x) {
```
Comment (pcwalton) [inline components/gfx/paint_context.rs]: I like the use of pattern matching, but this seems simpler as an if and then another nested if.

File: components/gfx/paint_context.rs (hunk 3 above)
Comment (bjwbell) [inline components/gfx/paint_context.rs]: I'd think, I'd have to add duplicated code in the 

    let ((box_TL, inner_TL), (box_TR, inner_TR), ...) = PaintContext::border_bounds(...) 

assignments to calculate `box_TL, inner_TL, box_TR, inner_TR`, if I changed it to two Rects.

Comment (pcwalton) [inline components/gfx/paint_context.rs]: I'm confused -- there should be `bottom_right()`, etc, methods on Rect, no?

Comment (bjwbell) [inline components/gfx/paint_context.rs]: yes, that's what I meant by look at the docs :-)

File: components/gfx/paint_context.rs (hunk 4 above)
Comment (bjwbell) [inline components/gfx/paint_context.rs]: I didn't think of it, changed

File: components/gfx/paint_context.rs [hunk 6]
```diff
@@ -466,6 +466,156 @@ impl<'a> PaintContext<'a> {
...
+    }
+
+    #[allow(non_snake_case)]
+    fn corner_bounds(bounds: &Rect<f32>,
+                     border: &SideOffsets2D<f32>,
+                     radii: &BorderRadii<AzFloat>) -> (Point2D<f32>,
+                                                       Point2D<f32>,
+                                                       Point2D<f32>,
+                                                       Point2D<f32>,
+                                                       SideOffsets2D<Size2D<f32>>) {
```
Comment (pcwalton) [inline components/gfx/paint_context.rs]: Could you return a rect here in lieu of the first four tuple elements?

**Comment by pcwalton** [comment]
Just one more comment and this looks great! Thanks!

File: components/gfx/paint_context.rs (hunk 6 above)
Comment (bjwbell) [inline components/gfx/paint_context.rs]: The tuples don't form a rect. That's why I had to do commit https://github.com/bjwbell/servo/commit/5819a2756d045cb066228b96e5a9a2050c1c16ae. 

I would like to something beside four tuples, I'm just not sure what. 
I'll have a followup PR for adding shorthand elliptical radius parsing e.g. `border-radius: 10px 5% / 20px`. 
I could fixup the four tuples then.

Comment (pcwalton) [inline components/gfx/paint_context.rs]: Oh, I see, right. Maybe just have a custom `struct` for this purpose? Something like:

    struct BorderOrigin {
        top_left: Point2D<f32>,
        ...
    }

Comment (bjwbell) [inline components/gfx/paint_context.rs]: That'll work. I'll go ahead and add it.

**Comment by bjwbell** [comment]
@pcwalton do you want me to squash?

**Comment by pcwalton** [comment]
Please do and then I'll r+ :)

**Pull request merged**